util/src/main/java/io/kubernetes/client/util/WebSocketStreamHandler.java

```diff
@@ -38,76 +39,97 @@
 public class WebSocketStreamHandler implements WebSockets.SocketListener, Closeable {
   private static final Logger log = LoggerFactory.getLogger(WebSocketStreamHandler.class);
 
-  Map<Integer, PipedOutputStream> output;
-  Map<Integer, PipedInputStream> input;
-  WebSocket socket;
-  String protocol;
+  private final Map<Integer, PipedInputStream> input = new HashMap<>();
+  private final Map<Integer, PipedOutputStream> pipedOutput = new HashMap<>();
+  private final Map<Integer, OutputStream> output = new HashMap<>();
+  private WebSocket socket;
 
-  public WebSocketStreamHandler() {
-    output = new HashMap<>();
-    input = new HashMap<>();
-  }
+  @SuppressWarnings("unused")
+  private String protocol;
+
+  private State state = State.UNINITIALIZED;
+
+  private static enum State {
+    UNINITIALIZED,
+    OPEN,
+    CLOSED
+  };
 
   @Override
-  public void open(String protocol, WebSocket socket) {
+  public synchronized void open(String protocol, WebSocket socket) {
+    if (state != State.UNINITIALIZED) throw new IllegalStateException();
     this.protocol = protocol;
     this.socket = socket;
+    state = State.OPEN;
+    notifyAll();
   }
 
   @Override
   public void bytesMessage(InputStream in) {
     try {
-      OutputStream out = getSocketInputOutputStream(in.read());
-      ByteStreams.copy(in, out);
+      handleMessage(in.read(), in);
     } catch (IOException ex) {
-      log.error("Error writing message", ex);
+      log.error("Error reading message channel", ex);
     }
   }
 
   @Override
   public void textMessage(Reader in) {
     try {
-      OutputStream out = getSocketInputOutputStream(in.read());
-      InputStream inStream =
-          new ByteArrayInputStream(CharStreams.toString(in).getBytes(Charsets.UTF_8));
-      ByteStreams.copy(inStream, out);
+      handleMessage(
+          in.read(), new ByteArrayInputStream(CharStreams.toString(in).getBytes(Charsets.UTF_8)));
     } catch (IOException ex) {
       log.error("Error writing message", ex);
     }
   }
 
+  protected void handleMessage(int stream, InputStream inStream) {
+    try {
+      OutputStream out = getSocketInputOutputStream(stream);
+      ByteStreams.copy(inStream, out);
+    } catch (IOException ex) {

```

Comment:
Given that both callers of `handleMessage` still catch `IOException` I would just have `handleMessage` throw IOException.